File: classe/gincana.c
#include <stdio.h>
#include <stdlib.h>


int main(){
    
    int opt;
    // Votos - meninos (xerife)
    int Joao = 0, Pedro = 0, Leandro = 0;
    // Votos - meninas (sinhazinha)
    int Maria = 0, Julia = 0, Bruna = 0;
    int qtd; 
    int voto;

    //Menu de áreas
    while (1)
    {
        printf("\nMENU\n");
        printf("[1] Explicacao da gincana\n");
        printf("[2] Comprar cartela \n");
        printf("[3] Ver ganhadores\n");
        printf("[0] Sair \n");
        printf("Escolha uma opcao: ");
        scanf("%d", &opt);

        if(opt == 0) {
            printf("Saindo...\n");
            break;
        }

        // Área de exploração da gincana
        switch (opt) {
            case 1:
            printf("\n\n EXPLICACAO DA GINCANA \n\n");
            printf("E ai, pessoal! Preparem os chapeus e os vestidos de chita! \n");
            printf("A nossa tradicional Gincana Junina de Xerife e Sinhazinha esta de volta e a competicao vai ser quente!\n");
            printf("Este ano, a disputa pelo titulo mais cobicado da festa envolve nao so carisma, mas tambem muita estrategia!\n\n");

            printf("Como funciona? E simples e divertido!\n");
            printf("Os candidatos a Xerife e Sinhazinha estao com suas cartelas de voto a postos.\n");
            printf("Cada cartela vendida conta como um ponto para o seu candidato favorito.\n");
            printf("O objetivo e mobilizar a galera, vender o maximo de cartelas possivel e garantir a vitoria!\n\n");

            printf("- Cada cartela = um voto!\n");
            printf("- O candidato (ou dupla) que acumular a maior quantidade de cartelas vendidas ate o dia da apuracao\n");
            printf("  sera coroado o GRANDE VENCEDOR da nossa festa junina! \n\n");
            break;
        
            //Área de compra de cartelas 
            case 2:{
                int categoria;

            printf("\n == Categoria ==\n");
            printf("[1] Votar no XERIFE (meninos)\n");
            printf("[2] Votar na SINHAZINHA (meninas)\n");
            printf("Escolha a categoria:");
            scanf(" %d", &categoria);

            if (categoria != 1 && categoria !=2){
                printf("Categoria invalida. Tente novamente.\n");
                break;
            }
            if (categoria == 1){
                printf("\n\n Candidatos a xerife \n\n");
                printf("1 - Joao\n");
                printf("2 - Pedro\n");
                printf("3 - Leandro\n");
                printf("Digite o numero do candidato escolhido: ");
                scanf(" %d", &voto);

            printf("Quantas cartelas voce deseja comprar? ");
            scanf("%d", &qtd);

            switch (voto){
                case 1: Joao += qtd; break;
                case 2: Pedro += qtd; break;
                case 3: Leandro += qtd; break;
                default:(printf("Inválido!\n"));
            }
            }
            else if(categoria == 2){
                printf("\n\n Candidatos a sinhazinha\n\n");
                printf("4 - Maria\n");
                printf("5 - Julia\n");
                printf("6 - Bruna\n");
                printf("Digite o numero do canditado escolhido: ");
                scanf(" %d", &voto);

            printf("Quantas cartelas voce deseja comprar? ");
            scanf("%d", &qtd);

            switch (voto){
                case 4: Maria += qtd; break;
                case 5: Julia += qtd; break;
                case 6: Bruna += qtd; break;
                default:
                    printf("Invalido!\n");
            }
        }
            else {
                printf("Categoria invalida. Tente novamente.\n");
            }
            printf("Voto registrado!\n");
            break;

            //Área feita para mostrar os ganhadores
            case 3:
            printf("\n RESULTADOS  \n");

            printf("\n--XERIFE (meninos)--\n");
            printf("Joao: %d votos\n", Joao);
            printf("Pedro: %d votos\n", Pedro);
            printf("Leandro: %d votos\n", Leandro);

            printf("\n--SINHAZINHA (meninas)--\n");
            printf("Maria: %d votos\n", Maria);
            printf("Julia: %d votos\n", Julia);
            printf("Bruna: %d votos\n", Bruna);

            int maiorX = Joao;
    
            if (Pedro > maiorX) maiorX = Pedro;
            if (Leandro > maiorX) maiorX = Leandro;

            int maiorS = Maria;
            if (Julia > maiorS) maiorS = Julia;
            if (Bruna > maiorS) maiorS = Bruna;

            printf("\nGanhadores:\n");
            printf("\nXERIFE:\n ");
            if (Joao== maiorX) printf("Joao com %d votos!\n", maiorX);
            if (Pedro== maiorX) printf("Pedro com %d votos!\n", maiorX);
            if (Leandro== maiorX) printf("Leandro com %d votos!\n", maiorX);

            printf("\nSINHAZINHA:\n ");
            if (Maria== maiorS) printf("Maria com %d votos!\n", maiorS);
            if (Julia== maiorS) printf("Julia com %d votos!\n", maiorS);
            if (Bruna== maiorS) printf("Bruna com %d votos!\n", maiorS);
            break;

            case 0: 
            printf("Encerrando o programa...\n");
            exit(0);

            default:
            printf("Opcao invalida.\n");
            }
        }
    }
    return 0;
}

File: classe/gincana.py
#Criação da classe candidato
class Candidato:
    def __init__(self, nome):
        self.nome = nome
        self.votos = 0

    def adicionar_votos(self, qtd):
        self.votos += qtd


class Categoria:
    def __init__(self, nome, candidatos):
        self.nome = nome
        self.candidatos = candidatos

    def mostrar_candidatos(self):
        print(f"\n--- Candidatos a {self.nome} ---")
        for i, cand in enumerate(self.candidatos, start=1):
            print(f"[{i}] {cand.nome} ({cand.votos} votos)")

#Sistema de quantidade de votos
    def votar(self):
        self.mostrar_candidatos()
        op = int(input("Digite o número do candidato escolhido: "))

        if op < 1 or op > len(self.candidatos):
            print("Opção inválida!")
            return

        qtd = int(input("Quantas cartelas deseja comprar? "))
        self.candidatos[op - 1].adicionar_votos(qtd)
        print("Voto registrado!\n")

#Aqui ele compara os votos dos candidatos e retorna o maior
    def vencedor(self):
        maior = max(self.candidatos, key=lambda c: c.votos)
        return maior

#Separação entre candidatos meninos e meninas
class Gincana:
    def __init__(self):
        self.xerife = Categoria("XERIFE (meninos)", [
            Candidato("João"),
            Candidato("Pedro"),
            Candidato("Leandro")
        ])

        self.sinhazinha = Categoria("SINHAZINHA (meninas)", [
            Candidato("Maria"),
            Candidato("Júlia"),
            Candidato("Bruna")
        ])

    def explicar(self):
        print("\n\nEXPLICAÇÃO DA GINCANA\n")
        print("E aí, pessoal! Preparem os chapéus e os vestidos de chita! 🤠💃")
        print("A nossa tradicional Gincana Junina de Xerife e Sinhazinha está de volta!")
        print("Cada cartela comprada soma um ponto para o candidato escolhido.")
        print("Quem vender mais cartelas será o grande vencedor!\n\n")

    def comprar_cartelas(self):
        print("\nCategorias para votar:\n")
        print("[1] Xerife (meninos)")
        print("[2] Sinhazinha (meninas)")
        categoria = int(input("Escolha a categoria: "))

        if categoria == 1:
            self.xerife.votar()
        elif categoria == 2:
            self.sinhazinha.votar()
        else:
            print("Categoria inválida!")

    def mostrar_resultados(self):
        print("\n----- RESULTADOS -----\n")

        print("\n-- XERIFE --")
        for c in self.xerife.candidatos:
            print(f"{c.nome}: {c.votos} votos")
        vencedor_x = self.xerife.vencedor()
        print(f"👉 Novo Xerife: {vencedor_x.nome} com {vencedor_x.votos} votos!")

        print("\n-- SINHAZINHA --")
        for c in self.sinhazinha.candidatos:
            print(f"{c.nome}: {c.votos} votos")
        vencedor_s = self.sinhazinha.vencedor()
        print(f"🌸 Nova Sinhazinha: {vencedor_s.nome} com {vencedor_s.votos} votos!\n")

    def menu(self):
        while True:
            print("\n----- MENU -----")
            print("[1] Explicação da gincana")
            print("[2] Comprar cartela")
            print("[3] Ver ganhadores")
            print("[0] Sair")
            op = int(input("Escolha uma opção: "))

            if op == 1:
                self.explicar()
            elif op == 2:
                self.comprar_cartelas()
            elif op == 3:
                self.mostrar_resultados()
            elif op == 0:
                print("Saindo...")
                break
            else:
                print("Opção inválida!\n")

#Inicia o sistema
g = Gincana()
g.menu()